### PR TITLE
Add support for Versions Catalogs `gradle/libs.versions.toml`

### LINKED ISSUE #333: Add support for Gradle 7.4 Version Catalogs
Gradle 7.0 introduced the "[centralized declaration of dependencies](https://docs.gradle.org/7.0/release-notes.html#centralized-dependency-versions)".
Namely, dependencies can be aliased, grouped, and versioned in a TOML file.
It would be incredibly nice if refreshVersions could target the `libs.versions.toml` file in future, rather than relying on a custom `version.properties`.

## Update: notes from jmfayard

the plan for now is not to give up `versions.properties`

it's to auto-generate the libraries versions catalog, with the placeholder version `_` and use the typesafe accessors to help migrate an existing project to refreshVersions

Running `$ ./gradlew refreshVersionsCatalog` would

   - Enable catalog version if necessary

```
// Enable Gradle's version catalog support
// https://docs.gradle.org/current/userguide/platforms.html
enableFeaturePreview("VERSION_CATALOGS")
```

   - Generate versions.properties, with all version keys (marked as unused)
   - Generate `gradle/libs.dependencies.toml`
   - Use the IDE and auto-completion to replace hardcoded dependencies with refreshVerion's dependency notations and versionCatalog or buildSrc/Libs.kt
   - To help with this, this task will show which typesafe accessors are available `$ ./ gradlew refreshVersionsMapping`

```
androidx.car:car                             | AndroidX.car
com.retrofit:retrofit-converter-json          | Square.retrofit.converter.json
org.koin:koin                                  |  libs.koin
```

Documentation

- [Sharing dependency versions between projects](https://docs.gradle.org/7.0-rc-1/userguide/platforms.html#sub:central-declaration-of-dependencies)
- [Cédric Champeau's blog: Simplified version management with Gradle 7](https://melix.github.io/blog/2021/03/version-catalogs.html)
- [tivi/libs.versions.toml at 4e378d36d37a878060bc824ab63adcc5408b203b · chrisbanes/tivi](https://github.com/chrisbanes/tivi/blob/4e378d36d37a878060bc824ab63adcc5408b203b/gradle/libs.versions.toml)

### PR DESCRIPTION
**Pull request by jmfayard**
Hello & bonjour,

## What?

Resolves #333

Add support for [Versions Catalogs](https://docs.gradle.org/current/userguide/platforms.html) `gradle/libs.versions.toml` in all our tasks (refreshVersions, refreshVersionsMigrate, refreshVersionsCleanup, refreshVersionsCatalog)

## Why?

First class support for Gradle versions catalogs, removing fear of plugin lock-in for our users.

## Testing?

You can test it in your own project, because I published a snapshot version on Sonatype

Setup refreshVersions `"0.41.0-SNAPSHOT"` as in https://jmfayard.github.io/refreshVersions/setup/#if-you-want-to-use-a-snapshot-version

Then:

```bash
# Versions Catalog are stable in Gradle 7.4
$ ./gradlew wrapper --gradle-version 7.4

# Generate the versions catalog
$ ./gradlew -refresh-dependencies refreshVersionsCatalog --versions --all
        modified:   versions.properties
        modified:   gradle/libs.versions.toml

# Migrate your build.gradle(.kts)
$ ./gradlew refreshVersionsMigrate --toml
        modified:   /Users/jmfayard/tignum/backend-tignum-x/build.gradle.kts

# Show dependency updates in gradle/libs.versions.toml
$ ./gradlew refreshVersions
        modified:   versions.properties
        modified:   gradle/libs.versions.toml
```

## How?

- `./gradlew refreshVersionsCatalog` generate `gradle/libs.versions.toml`
   - [x] only for Gradle >= 7.4
   - [x] with placeholder version `"_"`
   - [x] with hardcoded version when `--versions` option is used
   - [x] only for dependencies without a built-in dependency notation
   - [x] except if option `--all` is used
   - [x] adds plugins in the `[plugins]` section
   - [x] should be incremental: don't overwrite an existing entry
   - [x] should support refreshVersions dependency rules, create a `versions` entry and use `version.ref` for the dependency
   - [x] the `versions` section should be on top
   - [x] should support configuration cache https://github.com/jmfayard/refreshVersions/pull/509#issuecomment-1051381827
- `./gradlew refreshVersionsMigrate` 
  - [x] should support dependencies from `gradle/libs.versions.toml`
  - [x] option `--toml` should migrate to dependencies from `libs.versions.toml` before built in dependencies
  - [x] don't migrate `build.gradle` files inside a `resources` or `build` directory
- `./gradlew refreshVersions`
   - [x] should ignore entries from the versions catalog using placeholder version `"_"`
   - [x] should update entries with a hardcoded version
   - [x] should update entries using `versions.ref = "xxx"`
   - [x] should be idempotent
   - [x] add unit tests
   - [x] should not warn of hardcoded dependency if it comes from `gradle/libs.versions.toml`
   - [x] supports custom `projet.buildFileName=xxx", see https://github.com/jmfayard/refreshVersions/pull/509#issuecomment-1056935799
- [x] `./gradlew refreshVersionsCleanup` should remove comments in `gradle/libs.versions.toml`
- [x] Update `sample-kotlin`
- [ ] Documentation website should be updated
- [x] Publish a version `"0.41.0-SNAPSHOT"`
- [x] Ask for feedback on the issue and on Slack
- [ ] Write blog post about it
- [ ] `@LouisCad` to review it when it's not draft anymore

### CODE REVIEW

**Comment by yogurtearl** [comment]
Cool! Tried out the snapshot. :) 

it inlines all the versions into `[libraries]`, but also leaves the old versions unmodified under `[versions]`
```
[versions]
slf4j = "1.7.30"
log4j2 = "2.17.1"
kotlinLogging = "2.0.10"

[libraries]
slf4j-api = { module = "org.slf4j:slf4j-api", version.ref = "slf4j" }
slf4j-log4j12 = { module = "org.apache.logging.log4j:log4j-slf4j-impl", version.ref = "log4j2" }
log4j2 = { module = "org.apache.logging.log4j:log4j-core", version.ref = "log4j2" }
kotlin-logging = { module = "io.github.microutils:kotlin-logging-jvm", version.ref = "kotlinLogging" }

[bundles]
kotlinSlf4jLog4j2 = ["slf4j-api", "slf4j-log4j12", "log4j2", "kotlin-logging"]
```

if I run `./gradlew refreshVersionsCatalog --versions --all --no-configuration-cache` 

it outputs something like this: 
```
[versions]
slf4j = "1.7.29"
log4j2 = "2.17.1"
kotlinLogging = "2.0.9"

[libraries]

kotlin-logging-jvm = "io.github.microutils:kotlin-logging-jvm:2.0.10"

log4j-core = "org.apache.logging.log4j:log4j-core:2.17.1"

log4j-slf4j-impl = "org.apache.logging.log4j:log4j-slf4j-impl:2.17.1"

slf4j-api = "org.slf4j:slf4j-api:1.7.30"

[bundles]
kotlinSlf4jLog4j2 = ["slf4j-api", "slf4j-log4j12", "log4j2", "kotlin-logging"]
```

Now we have duplicated versions and the `kotlinSlf4jLog4j2` bundle reference a `slf4j-log4j12` library that no longer exists.
Also, the next build fails with: 
```
* What went wrong:
org.gradle.api.InvalidUserDataException: Invalid catalog definition:
  - Problem: In version catalog libs, a bundle with name 'kotlinSlf4jLog4j2' declares a dependency on 'slf4j.log4j12' which doesn't exist.
    
    Reason: Bundles can only contain references to existing library aliases.
    
    Possible solutions:
      1. Make sure that the library alias 'slf4j.log4j12' is declared.
      2. Remove 'slf4j.log4j12' from bundle 'kotlinSlf4jLog4j2'.
    
    Please refer to https://docs.gradle.org/7.4/userguide/version_catalog_problems.html#undefined_alias_reference for more details about this problem.
```

Maybe it should remove the `[versions]` section and update the bundles with the new names?

**Comment by yogurtearl** [comment]
with something like this: 

```
[versions]
kotlin = "1.6.0"

[plugins]
kotlin-gradle = { id = "org.jetbrains.kotlin.jvm", version.ref = "kotlin" }
```

it seems to just ignore the `[plugins]` block.

**Comment by yogurtearl** [comment]
Also, it doesn't seem to work nicely with Configuration cache. 
https://docs.gradle.org/7.4/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution

```
invocation of Task.project📋 at execution time is unsupported.[ ?](https://docs.gradle.org/7.4/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution)
⌄task:refreshVersionsCatalog📋 of type de.fayard.refreshVersions.RefreshVersionsCatalogTask📋
⌄exception stack trace 📋
org.gradle.api.InvalidUserCodeException: Invocation of 'Task.project' by task ':refreshVersionsCatalog' at execution time is unsupported.
	at org.gradle.configurationcache.initialization.DefaultConfigurationCacheProblemsListener.onTaskExecutionAccessProblem(ConfigurationCacheProblemsListener.kt:74)
	at org.gradle.configurationcache.initialization.DefaultConfigurationCacheProblemsListener.onProjectAccess(ConfigurationCacheProblemsListener.kt:53)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:464)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:446)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:61)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:434)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast.dispatch(DefaultListenerManager.java:221)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast.dispatch(DefaultListenerManager.java:192)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy170.onProjectAccess(Unknown Source)
	at org.gradle.api.internal.AbstractTask.notifyProjectAccess(AbstractTask.java:1053)
	at org.gradle.api.internal.AbstractTask.getProject(AbstractTask.java:237)
	at org.gradle.api.DefaultTask.getProject(DefaultTask.java:59)
	at de.fayard.refreshVersions.RefreshVersionsCatalogTask.refreshVersionsCatalogAction(RefreshVersionsCatalogTask.kt:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
```

**Comment by yogurtearl** [comment]
If I have something like this, it seems to just remove these libraries altogether. 
These are only used in my `buildSrc` dir. 

```
[libraries]
kotlin-gradle = { module = "org.jetbrains.kotlin:kotlin-gradle-plugin", version.ref = "kotlin" }
detekt-gradle = { module = "io.gitlab.arturbosch.detekt:detekt-gradle-plugin", version.ref = "detekt" }
```

**Comment by yogurtearl** [comment]
I have this in `buildSrc/settings.gradle.kts` to enable using the same version catalog in `buildSrc` and in my build: 

```
dependencyResolutionManagement {
    versionCatalogs {
        create("libs") {
            from(files("../gradle/libs.versions.toml"))
        }
    }
}
```

and I use some of the `[libraries]` in `buildSrc/build.gradle.kts`

after I run 
`./gradlew refreshVersionsCatalog --versions --all --no-configuration-cache`

when I run: 
`./gradlew refreshVersionsMigrate --no-configuration-cache`

if fails because `buildSrc/build.gradle.kts` is referencing the old library names.

**Comment by jmfayard** [comment]
@yogurtearl thanks a lot for trying out the snapshot!

I though initially the task `./gradlew refreshVersionsCatalog` as a one-off thing that you use when you don't have versions catalog and then never use again. So yes, currently the `libraries` section is rewriting from scratch and the other are ignored. Which indeed create a build error next time if you already use the versions catalog. I will improve this.

Thanks for your reports on the configuration cache, I will have a look.

**Comment by jmfayard** [comment]
Turns out you have to do `./gradlew --refresh-dependencies refreshVersionsCatalog --all --versions`  to refresh the snapshot dependency

**Comment by runningcode** [comment]
I tried out the snapshot. Works quite well for the most part.

Things that did not work:
It removed the versions from the `plugins` block in an included build and the main build.

The issue is that the included build doesn't use the toml file. Even after adding the toml file to the included build, the plugins block was not replaced with `alias(libs.plugin.mydependency)`.

In the main build plugin versions also lost their plugin id versions instead of being replaced with `alias`.

For testing dependencies, such as junit, their declarations were replaced with `Testing.junit4` instead of `libs.junit.junit`.

**Comment by jmfayard** [comment]
Thanks @runningcode 
I now ignore the  `build.gradle` files from the resoruces/build directory and added an option `--toml` so that `refreshVersionsMigrate` pick the dependencies from libs.versions.toml first

**Comment by LouisCAD** [comment]
Thanks all for your feedback, and of course Jean-Michel for leading this!

I will take the time to review this early next week.

**Comment by jmfayard** [comment]
@Dave562CZ Can you put this snippet in the root `./build.gradle.kts` and tell me if it correctly finds all your build files?
`project.allprojects { println("buildFile=$buildFile") }`

**Comment by jmfayard** [comment]
@yogurtearl it turns out the tasks from refreshVersions cannot be compatible with the configuration cache for now (we need access to the dependency graph itself)

But I now activated `Task.notCompatibleWithConfigurationCache()`

---

[Declare a task incompatible with the configuration cache](https://docs.gradle.org/7.4/userguide/configuration_cache.html#config_cache:task_opt_out)

It is possible to declare that a particular task is not compatible with the configuration cache via the [Task.notCompatibleWithConfigurationCache()](https://docs.gradle.org/7.4/javadoc/org/gradle/api/Task.html#notCompatibleWithConfigurationCache-java.lang.String-) method.

Configuration cache problems found in tasks marked incompatible will no longer cause the build to fail.

**Comment by yogurtearl** [comment]
When I add this to a project that already uses `lib.versions.toml`, but doesn't use `refreshVersions`.
it still creates a `versions.properties` file. 🤔  Is that the expected behavour? 
Would like to have only `lib.versions.toml` and no `versions.properties` and have the `refreshVersions` plugin update the versions in `lib.versions.toml`

**Comment by yogurtearl** [comment]
if I have: 
```
settings.gradle.kts
features
└── feature1
 │  └── build.gradle.kts
└── feature1
    └── build.gradle.kts
```

When I run `./gradlew refreshVersionsMigrate --toml`

it fails with 
```
* What went wrong:
Execution failed for task ':refreshVersionsMigrate'.
> features/build.gradle (No such file or directory)
```

there may be sub-directories that don't have a `build.gradle(.kts)`

**Comment by LouisCAD** [comment]
> When I run `refreshVersionsCatalog` to a project that already uses `lib.versions.toml`, but doesn't use `refreshVersions`. it still creates a `versions.properties` file. 🤔  Is that the expected behavour? 
> 
> Would like to have only `lib.versions.toml` and no `versions.properties` and have the `refreshVersions` plugin update the versions in `lib.versions.toml`
> 

Yes, it's expected, we are not really handling full version catalog yet.

**Comment by Dave562CZ** [comment]
> @Dave562CZ Can you put this snippet in the root `./build.gradle.kts` and tell me if it correctly finds all your build files? `project.allprojects { println("buildFile=$buildFile") }`

yes it does. I actually used same API for renaming files back to build.gradle.kts apply migration and renaming back to module names, it worked this way

**Comment by NikkyAI** [comment]
from what i can see currently single dependencies that do not have `extraArtifactVersionKeyRules` defined for them are now mixed into the `[libraries]` section

i think it would make more sense to generate a kay for them just like refreshVersion used to do for the `version.properties` file
it might bloat the `libs.version.toml` but i think having all versions in the versions section is worth it
as currently i end up with some depdencies in random places in the toml file
while i would like to just fold that section closed and let `refreshVersions` deal with it

also it seems like for gradle to use the plugins from version catalog.. the plugins block need to look like so:

```kotlin
plugins {
    alias(libs.plugins.org.jetbrains.kotlin.jvm)
    alias(libs.plugins.org.jetbrains.kotlin.plugin.serialization)
    alias(libs.plugins.com.google.devtools.ksp)
    alias(libs.plugins.com.github.johnrengelman.shadow)
    alias(libs.plugins.com.squareup.sqldelight)
    application
}
```

about the `[plugins]` section in the toml.. same comment about mixing versions and version references..
would be nice to have it moved below the `[versions]`

and.. will i be able to continue using `group:artifact:_` syntax in my buildscripts and have refreshVersion add them to the toml instead of `versions.properties` and lookup (or change it under the hook, idc) from version catalog ?
right now i cannot leave the `versions.properties` empty as it just gets repopulated

one of the biggest selling points of `refreshVersions` for me was to just be able to drop in some dependency, have refreshVersions grab a sensible default and list newer available versions

seems like adding dependencies has become a bit more tedious when i have to add it to the `[libraries]` and `[versions]` section and finally to the buildscript

**Comment by NikkyAI** [comment]
i also noticed that the libraries do not generate very useful identifiers right now..

migrating `implementation("dev.kord.x:emoji:_")`

with a rule
```
dev.kord.x:emoji
    ^^^^^^.^^^^^
```

i expected 

```toml
[versions]

kord-x-emoji = "0.5.0"

[libraries]

kord-x-emoji = { group = "dev.kord.x", name = "emoji", version.ref = "kord-x-emoji" }
```

but got

```toml
[versions]

kord-x-emoji = "0.5.0"

[libraries]

emoji = { group = "dev.kord.x", name = "emoji", version.ref = "kord-x-emoji" }
```

using only the `name` component of the libraries key seems like it would lead to not very clear dependency notation in the buildscripts
and most likely conflicts too

eg:
```
implementation(libs.annotations) // it it jetbains annotation or `com.kotlindiscord.kord.extensions:annotations` ?
```

better would be to use `"${version.ref}-$name"` as a more.. unique identifier

```toml
kord-extensions-annotations = { group = "com.kotlindiscord.kord.extensions", name = "annotations", version.ref = "kord-extensions" }
```

```kts
implementation(libs.kord.extensions.annotations)
```

**Comment by jmfayard** [comment]
@NikkyAI if you have two different dependencies called `annnotations`, I will detect it and use longer names

Regarding [your first message](https://github.com/jmfayard/refreshVersions/pull/509#issuecomment-1059070441):

If that's how you want to use refreshVersions, I would suggest you keep all versions in `versions.properties`
and use the placeholder version `_`  inside `gradle/libs.versions.toml`

You do that with `./gradlew refreshVersionsCatalog` instead of `./gradlew refreshVersionsCatalog --versions --all`

That's like a more modern version of `buildSrcLibs`

To clarify: we don't drop support for `versions.properties`, what we do is that add support for `gradle/libs.versions.toml`

**Comment by tbroyer** [comment]
Trying it out on a project (using exclusively a version catalog, I'm looking for a plugin that would help me find newer versions of the plugins and libs), I get deprecation warnings:
```
> Task :refreshVersions FAILED
The VersionCatalog.getDependencyAliases() method has been deprecated. This is scheduled to be removed in Gradle 8.0. Please use the getLibraryAliases() method instead. Consult the upgrading guide for further information: https://docs.gradle.org/7.4.2/userguide/upgrading_version_7.html#version_catalog_deprecations
The VersionCatalog.findDependency(String) method has been deprecated. This is scheduled to be removed in Gradle 8.0. Please use the findLibrary(String) method instead. Consult the upgrading guide for further information: https://docs.gradle.org/7.4.2/userguide/upgrading_version_7.html#version_catalog_deprecations
```
and an NPE:
```
Caused by: java.lang.NullPointerException
        at de.fayard.refreshVersions.core.internal.TomlLine.getVersion(TomlLine.kt:35)
        at de.fayard.refreshVersions.core.internal.TomlUpdater.linesForUpdate(TomlUpdater.kt:60)
        at de.fayard.refreshVersions.core.internal.TomlUpdater.updateNewVersions(TomlUpdater.kt:41)
        at de.fayard.refreshVersions.core.internal.TomlUpdater.updateNewVersions(TomlUpdater.kt:14)
        at de.fayard.refreshVersions.core.RefreshVersionsTask$taskActionRefreshVersions$1.invokeSuspend(RefreshVersionsTask.kt:85)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:279)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at de.fayard.refreshVersions.core.RefreshVersionsTask.taskActionRefreshVersions(RefreshVersionsTask.kt:62)
```

**Comment by jmfayard** [comment]
@tbroyer thanks a lot for the bug report.
Dependencies using a BOM/platform are now supported by all three commands : `refreshVersions`, `refreshVersionsMigrate`, `refreshVersionsCatalog`

**Comment by joffrey-bion** [comment]
Thanks a lot for this PR, this would be a reason for me to adopt this plugin. It's the first time I'm trying it. I have been using the versions catalog for a while and wanted to try out this plugin to find new versions to upgrade more easily.

Trying `0.40.2-SNAPSHOT`, when running `./gradlew refreshVersions` it seems ALL the available versions of the same major version of each dependency are listed in my `[versions]` section:

```
[versions]
jackson = "2.13.2"
```
Is changed to:
```
## Generated by $ ./gradlew refreshVersionsCatalog

[versions]

jackson = "2.13.2"
### available = "2.0.0-RC1"
### available = "2.0.0-RC2"
### available = "2.0.0-RC3"
### available = "2.0.0"
### available = "2.0.1"
### available = "2.0.2"
### available = "2.0.4"
### available = "2.0.5"
### available = "2.0.6"
### available = "2.1.0"
### available = "2.1.1"
### available = "2.1.2"
### available = "2.1.3"
### available = "2.1.4"
### available = "2.1.5"
### available = "2.2.0-rc1"
### available = "2.2.0"
### available = "2.2.1"
### available = "2.2.2"
### available = "2.2.3"
### available = "2.2.4"
### available = "2.3.0-rc1"
### available = "2.3.0"
### available = "2.3.1"
### available = "2.3.2"
### available = "2.3.3"
### available = "2.3.4"
### available = "2.3.5"
### available = "2.4.0-rc1"
### available = "2.4.0-rc2"
### available = "2.4.0-rc3"
### available = "2.4.0"
### available = "2.4.1"
### available = "2.4.1.1"
### available = "2.4.2"
### available = "2.4.3"
### available = "2.4.4"
### available = "2.4.5"
### available = "2.4.6"
### available = "2.5.0-rc1"
### available = "2.5.0"
### available = "2.5.1"
### available = "2.5.2"
### available = "2.5.3"
### available = "2.5.5"
### available = "2.6.0-rc1"
### available = "2.6.0-rc2"
### available = "2.6.0-rc3"
### available = "2.6.0-rc4"
### available = "2.6.0"
### available = "2.6.1"
### available = "2.6.2"
### available = "2.6.3"
### available = "2.6.4"
### available = "2.6.5"
### available = "2.6.6"
### available = "2.6.7"
### available = "2.7.0-rc1"
### available = "2.7.0-rc2"
### available = "2.7.0-rc3"
### available = "2.7.0"
### available = "2.7.1"
### available = "2.7.2"
### available = "2.7.3"
### available = "2.7.4"
### available = "2.7.5"
### available = "2.7.6"
### available = "2.7.7"
### available = "2.7.8"
### available = "2.7.9"
### available = "2.8.0.rc1"
### available = "2.8.0.rc2"
### available = "2.8.0"
### available = "2.8.1"
### available = "2.8.2"
### available = "2.8.3"
### available = "2.8.4"
### available = "2.8.5"
### available = "2.8.6"
### available = "2.8.7"
### available = "2.8.8"
### available = "2.8.9"
### available = "2.8.10"
### available = "2.8.11"
### available = "2.9.0.pr1"
### available = "2.9.0.pr2"
### available = "2.9.0.pr3"
### available = "2.9.0.pr4"
### available = "2.9.0"
### available = "2.9.1"
### available = "2.9.2"
### available = "2.9.3"
### available = "2.9.4"
### available = "2.9.5"
### available = "2.9.6"
### available = "2.9.7"
### available = "2.9.8"
### available = "2.9.9"
### available = "2.9.10"
### available = "2.10.0.pr1"
### available = "2.10.0.pr2"
### available = "2.10.0.pr3"
### available = "2.10.0"
### available = "2.10.1"
### available = "2.10.2"
### available = "2.10.3"
### available = "2.10.4"
### available = "2.10.5"
### available = "2.11.0.rc1"
### available = "2.11.0"
### available = "2.11.1"
### available = "2.11.2"
### available = "2.11.3"
### available = "2.11.4"
### available = "2.12.0-rc1"
### available = "2.12.0-rc2"
### available = "2.12.0"
### available = "2.12.1"
### available = "2.12.2"
### available = "2.12.3"
### available = "2.12.4"
### available = "2.12.5"
### available = "2.12.6"
### available = "2.13.0-rc1"
### available = "2.13.0-rc2"
### available = "2.13.0"
### available = "2.13.1"
### available = "2.13.2"
### available = "2.13.3"
```

Is this expected?

Also, it generates a `versions.properties` anyway (with nothing inside but the header comment). Is there a way to disable this?

File: plugins/core/src/main/kotlin/de/fayard/refreshVersions/core/internal/VersionCatalogs.kt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import de.fayard.refreshVersions.core.internal.UsedPluginsHolder.ConfigurationLessDependency
+import org.gradle.api.artifacts.Dependency
+import org.gradle.api.artifacts.MinimalExternalModuleDependency
+import org.gradle.api.artifacts.VersionCatalog
+import org.gradle.util.GradleVersion
+
+@InternalRefreshVersionsApi
+object VersionCatalogs {
+
+    const val LIBS_VERSIONS_TOML = "gradle/libs.versions.toml"
```
Comment (yogurtearl) [inline plugins/core/src/main/kotlin/de/fayard/refreshVersions/core/internal/VersionCatalogs.kt:13]: the catalog file might not always be named `libs.versions.toml` and there might be multiple files. 

see https://sourcediving.com/manage-your-gradle-dependencies-with-version-catalog-not-only-in-android-114117647cdb#957a

Comment (joffrey-bion) [reply plugins/core/src/main/kotlin/de/fayard/refreshVersions/core/internal/VersionCatalogs.kt:13]: Agreed, although this might be good enough for a first version of the versions catalog support

**Comment by LouisCAD** [comment]
FYI, you can use the version `0.41.0-SNAPSHOT` now, it includes latest fixes merged from the main branch, and changes from this PR.

**Comment by LouisCAD** [comment]
Will be merged, one way or another through next week, or before.

Thank you all for your patience as I'm reviewing and polishing this 🙂.

**Comment by jmfayard** [comment]
@LouisCAD you renamed UsedPluginsHolder to UsedPluginsTracker
but there are still UsedPluginsHolder in the code like this one

https://github.com/jmfayard/refreshVersions/blob/bb241f39880e528335193574debdbf2d471f1c78/plugins/core/src/main/kotlin/de/fayard/refreshVersions/core/internal/DependencyNotations.kt#L84

I don't get how this can possibly compile on the CI, but it doesn't on my machine

**Comment by LouisCAD** [comment]
@jmfayard Can you try with a clean clone?

**Comment by joffrey-bion** [comment]
@LouisCAD any ETA on this? No rush of course, I just wanted to know where it was at, since your last comment suggested an imminent merge

**Comment by LouisCAD** [comment]
@joffrey-bion Hard to give an accurate, but I'll resume work on this on Wednesday, and I hope it's only 2 days of work away.

**Comment by LouisCAD** [comment]
@jmfayard We probably want tests for the `generateVersionsCatalogText` function (formerly named `versionsCatalog`.

If you're inspired, feel free to add at least one, and I'll complete it with edge cases like version-less dependencies (which are now handled).

**Comment by LouisCAD** [comment]
I can now give a more accurate ETA: Sometime next week, I'll make a release with this, probably behind a feature flag.

I just published another snapshot BTW with the latest changes fixing support for version-less dependencies when generating a version catalog with refreshVersions.

**Comment by jmfayard** [comment]
@LouisCAD I added tests for `refreshVersionsCatalog` and fixed a critical bug in the commit https://github.com/jmfayard/refreshVersions/pull/509/commits/d4c497bafd00cda4a605835497d5e3d6594772d9
`./gradlew refreshVersionsCatalog --versions` was generating the versions catalog with underscores instead of the real version, so people would not have been able to use the catalog without refreshVersions

**Comment by jmfayard** [comment]
I added the feature flag `VERSIONS_CATALOG` to minimize the risks of the release next week

**Comment by yschimke** [comment]
Just want to confirm _my_ use case for refreshVersions with version catalogs. See if it will be supported or adjust my thinking if not.

I'd like to use the typesafe refreshVersions dependency hierarchy in my build files.  But then have a version catalog in a form that tools like renovate can bump in PRs.  Would this work?

**Comment by jmfayard** [comment]
I don't understand

**Comment by yschimke** [comment]
@jmfayard I was trying to understand what version catalog support means for refreshVersions.

From the original feature request "It would be incredibly nice if refreshVersions could target the libs.versions.toml file in future, rather than relying on a custom version.properties."

But reading more it sounds like it's about migrating *to* refresh versions from version catalogs, and avoiding lockin.  I was hoping more for the original request.

If I look at the sample project in your PR https://github.com/jmfayard/refreshVersions/tree/7ebf046db1dfff642900911953b83588b6efc377/sample-kotlin

sample-kotlin.gradle.kts: dependencies declared in any of KotlinX.coroutines.core, com.google.guava:guava:15.0, com.google.guava:guava:15.0
versions.properties: dependencies defined in here like normal
libs.versions.toml: kept in sync with versions.properties, but generated each time.

Not a problem, just means https://github.com/jmfayard/refreshVersions/issues/207 would help me with renovate tooling.  Which is really nice.

**Comment by jmfayard** [comment]
@yschimke you will able to do many things, including using only the versions  catalogs, using only versions.properties, generating the versions catalog with refreshVersions and then removing refreshVersions, and a mix of all of that. I will publish an article once it's released, that will be clearer.

**Comment by yschimke** [comment]
Sounds impressive. And I'm really enjoying refresh versions as it is.

**Comment by LouisCAD** [comment]
@jmfayard

## ???????????????????!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Screenshot 2022-08-03 at 01 08 57

Screenshot 2022-08-03 at 01 06 58

**Comment by jmfayard** [comment]
@LouisCAD my bad, I never intended to force push on this branch.
The branch should be back in the right state

**Comment by LouisCAD** [comment]
Thanks for reverting that.

**Comment by LouisCAD** [comment]
The tests we have failed to catch an important issue:

The list of available updates depends on the contents of the `versions.properties` file.

Basically, we can end up with two sources of truth for one version, where both are lying.

I don't know how long it would take me to fix that, but the plan to release this week is maintained. That said, I'll flip the feature flag to disabled by default, and we will take the time to polish this feature before enabling it by default.

**Comment by LouisCAD** [comment]
The following malfunction examples have all been taken after running the `refreshVersions` task with the latest content of this branch.

## Example 1:

Screenshot 2022-08-04 at 01 54 50

## Example 2:

Screenshot 2022-08-04 at 01 55 43

## Example 3:

Screenshot 2022-08-04 at 01 59 45

**Comment by jmfayard** [comment]
I'm not sure I get it. We can have a call to discuss it if you want.

The source of truth is not any particular file (versions.properties / libs.versions.toml), it's what the Gradle API tells us is the version for dependencies that are actually used in the buiild. If you put a version in the libs;toml but don't use it the best we can do is add a `unused` comment.

**Comment by jmfayard** [comment]
In `sample-kotlin`, I did `gradle refreshVersionsCatalog --all --versions`
In `libs.versions.toml` I got

> `log4j-jul = "org.apache.logging.log4j:log4j-jul"`

When I refreshed Gradle, I got this error:

> `  - Problem: In version catalog libs, on alias 'log4j-jul' notation 'org.apache.logging.log4j:log4j-jul' is not a valid dependency notation.`

**Comment by LouisCAD** [comment]
Quick update:

I am a bit late on the ETA I provided, but we're almost done. Here's the proof that we're making progress:
IMAGE 2022-08-08 12:12:51

As you can see, we're now filtering the available versions properly, and we support usage of `versions.properties` and versions catalog together.

Note that we currently don't support non-default catalogs yet.

Two things left blocking this PR:
1. Migration of version-less dependencies
2. Disabling the feature by default while we continue testing remaining edge cases thoroughly

**Comment by LouisCAD** [comment]
I did further testing, and after my latest commits, it doesn't seem to introduce regressions, so I think we will keep the flag enabled by default.

**Comment by LouisCAD** [comment]
I don't know why the CI is failing, since it's giving 404 Not Found errors when I try to look…

I reported the issue to GitHub support.

On my machine the `check` task runs fine.

I was ready to merge this.

**Comment by LouisCAD** [comment]
GitHub Actions got fixed, and I added the missing opt-in that is not required as the Gradle 7.5.1 update updated Kotlin.

Now is time to merge!

**Comment by imashnake0** [comment]
pog
